frame/support/procedural/src/benchmark.rs

```diff
@@ -0,0 +1,816 @@
+					extrinsic_call,
+					verify_stmts: Vec::from(
+						&item_fn.block.stmts[(i + 1)..item_fn.block.stmts.len()],
+					),
+					extra,
+					skip_meta,
+				})
+			}
+			i += 1;
+		}
+		return Err(Error::new(
+			item_fn.block.brace_token.span,
+			"No valid #[extrinsic_call] annotation could be found in benchmark function body.",
+		))
+	}
+}
+
+/// Parses and expands a `#[benchmarks]` or `#[instance_benchmarks]` invocation
+pub fn benchmarks(attrs: TokenStream, tokens: TokenStream, instance: bool) -> TokenStream {
+	let module = parse_macro_input!(tokens as ItemMod);
+	let where_clause = match syn::parse::<Nothing>(attrs.clone()) {
+		Ok(_) => quote!(),
+		Err(_) => parse_macro_input!(attrs as WhereClause).predicates.to_token_stream(),
+	};
+	let mod_vis = module.vis;
+	let mod_name = module.ident;
+	let mut expanded_stmts: Vec<TokenStream2> = Vec::new();
+	let mut benchmark_defs: Vec<BenchmarkDef> = Vec::new();
+	let mut benchmark_names: Vec<Ident> = Vec::new();
+	let mut extra_benchmark_names: Vec<Ident> = Vec::new();
+	let mut skip_meta_benchmark_names: Vec<Ident> = Vec::new();
+	let mut args: Option<BenchmarkAttrs> = None;
+	if let Some(mut content) = module.content {
+		for stmt in &mut content.1 {
+			let mut found_item: Option<ItemFn> = None;
+			if let Item::Fn(func) = stmt {
+				let mut i = 0;
+				for attr in &func.attrs.clone() {
+					if let Some(seg) = attr.path.segments.last() {
+						if let Ok(_) =
+							syn::parse::<keywords::benchmark>(seg.ident.to_token_stream().into())
+						{
+							let tokens = attr.tokens.to_token_stream().into();
+							args = Some(parse_macro_input!(tokens as BenchmarkAttrs));
+							func.attrs.remove(i);
+							found_item = Some(func.clone());
+							i += 1;
+						}
+					}
+				}
+			}
+			if let (Some(item_fn), Some(args)) = (found_item, &args) {
+				// this item is a #[benchmark] or #[instance_benchmark]
+				let benchmark_def = match BenchmarkDef::from(&item_fn, args.extra, args.skip_meta) {
+					Ok(def) => def,
+					Err(err) => return err.to_compile_error().into(),
+				};
+
+				// expand benchmark
+				let expanded = expand_benchmark(
+					benchmark_def.clone(),
+					&item_fn.sig.ident,
+					instance,
+					where_clause.clone(),
+				);
+
+				// record benchmark name
+				let name = item_fn.sig.ident;
+				benchmark_names.push(name.clone());
+				if benchmark_def.extra {
+					extra_benchmark_names.push(name.clone());
+				}
+				if benchmark_def.skip_meta {
+					skip_meta_benchmark_names.push(name.clone())
+				}
+
+				expanded_stmts.push(expanded);
+				benchmark_defs.push(benchmark_def);
+			} else {
+				// this is not a benchmark item, copy it in verbatim
+				expanded_stmts.push(stmt.to_token_stream());
+			}
+		}
+	}
+
+	// generics
+	let generics = match instance {
+		false => quote!(T),
+		true => quote!(T, I),
+	};
+	let full_generics = match instance {
+		false => quote!(T: Config),
+		true => quote!(T: Config<I>, I: 'static),
+	};
```

Comment:
I think there are utility functions that you can use somewhere under frame support procedural.

Reply:
I poked around and couldn't find this. I did see some for `Storage` but not general ones for this situation as far as I can tell.